### DISCUSSION TITLE
Quinoa URL not available in prod-mode

### DISCUSSION THREAD

**Discussion by nimo23** [quarkiverse/quarkus-quinoa#804] [Q&A] [answered]
I have a **Jakarta Faces** app within `src/main/resources`:

![Screenshot at Nov 11 20-20-04](https://github.com/user-attachments/assets/3cf1ed70-1114-4418-9d69-a0af25d36ccb)

and a **Quinoa web app** (based on Vite) within `src/main/webui`. 
The `vite.config.ts`:

```
export default defineConfig({
    // application is deployed to the root domain
    base: '/',
    plugins: [react(), tsconfigPaths()],
    server: {
	port: 5173
    },
})
```

and the scripts in `package.json`:

```
"scripts": {
	"dev": "vite",
	"build": "tsc -b && vite build",
	"preview": "vite preview"
},
```

In **dev-mode**, it works well:
-  Going to `localhost:8080/index.xthml` (note the suffix `xhtml`) will serve **the jakarta faces app**.
- Going to `localhost:8080/index.html` (note the suffix `html`) will serve the **web app provided by quinoa**.

However, when building the Quarkus app (with `mvn clean package`), the **web app provided by Quinoa** is no longer available at the specified URL. If you go to `localhost:8080/index.html` (or `localhost:8080/index.xthml`), the **Jakarta Faces app** will be deployed. However, I want the same URL behavior as when running in `dev`-mode, i.e. going to `index.html` should serve **the web app provided by Quinoa** and going to to `index.xhtml`, the **Jakarta Faces app** should be served.

My Quinoa settings in `application.properties`:

```
quarkus.http.port=8080
quarkus.http.enable-compression=true
quarkus.quinoa.enable-spa-routing=true
```

The following assumption made in https://docs.quarkiverse.io/quarkus-quinoa/dev/index.html is also somewhat irritating to me:

> With Quinoa, you don’t need to manually copy the files to META-INF/resources. Quinoa has its own system and will provide another Vert.x route for it. **If you have conflicting files with META-INF/resources, Quinoa will have priority over them.**

This means that Quinoa's `webui` directory takes precedence over the contents of `META-INF/resources`, but in the case described above it seems that `META-INF/resources` takes precedence in prod mode. However, both are provided in `dev`-mode, which is what I want.

I guess I need a **custom configuration** to make it also work in `prod`-mode, i.e. serve **Quinoa web app** when going to `localhost:8080/index.html`, and serve J**akarta web app** when going to `localhost:8080/index.xhtml`. 
But what configuration do I have to do?

**Comment by melloware** [comment]
i am not sure i would try and combine these two things or you should serve your quinoa app at a different location.  I am actually surprised it works in Dev Mode

**Reply by nimo23** [reply]
Yes, it works smoothly in `Dev` mode and without additional settings. So it should also be possible to work in `Prod` mode. I suspect it's a Quarkus setting (or Quinoa setting), but I'm not sure which one. I also tried explicitly assigning `quarkus.http.cors=true`, but this not work. Hmm..

I also tried setting up another port for Quinoa (port 5173) and Jakarta Faces (port 8080) but that doesn't work either.

**Reply by melloware** [reply]
just to make sure did you crack open the `.jar` and make sure both files and everything is where you expect it in PRD mode inside the JAR?

**Reply by nimo23** [reply]
After the build (`mvn package`) there is the `quinoa`-folder (which contains the compiled webapp-contents) in the `target`-directory:

![Screenshot at Nov 11 20-53-40](https://github.com/user-attachments/assets/72215412-cd9b-4f7b-b0fd-16f283f81bf5)

So you mean that this content should also be in `src/main/resources` within `target/myApp.jar`? I looked in the (extracted) `jar`, there is no content from `webapp`-quinoa. But does it have to be in there? It's **not** a fat/uber-jar. And besides, if it has to be inside, why does not quinoa put these contents in? The Faces app content is still there (within `jar/META-INF/resources/`) and, I think, there are no conflicting files preventing Quinoa from placing this content in the required folder (if there is one).

**Reply by nimo23** [reply]
If Quinoa really needs to include the `webapp` content in the JAR file (which makes perfect sense), then I don't understand why it can't include that content. If this is not already possible (due to conflict files even if there are none in my app), it might be possible by providing another quinoa setting:

```
// stores the production build content 
// of webapp into the  jar/META-INF/resources/react
quarkus.quinoa.prod-dir=/react/
```

**Reply by nimo23** [reply]
I think, a new setting like `quarkus.quinoa.prod-dir` does not make sense as we already have:

```
// stores the production build content 
// of webapp into the  jar/META-INF/resources/react
quarkus.quinoa.build-dir: react
```

But nevertheless, quinoa does not package the webapp into the app's jar..

**Reply by melloware** [reply]
i have to test this but i actually think its a conflict of MyFaces and Quinoa.

**Reply by melloware** [reply]
I am pretty sure that is the issue is MyFaces Servlet is intercepting the requests BEFORE Quinoa gets them.
Reproducer:
[quinoa-faces.zip](https://github.com/user-attachments/files/17718819/quinoa-faces.zip)

Works fine in Dev mode but if you run..

```
mvn clean package -Dquarkus.package.jar.type=uber-jar
java -jar target/quinoa-faces-runner.jar
```

So in dev mode you can go to 

http://localhost:8080  -> Faces APP
http://localhost:8080/quinoa  -> Quinoa App

but in PROD mode if you do...

http://localhost:8080  -> Faces APP
http://localhost:8080/quinoa  -> 404 Page not found intercepted by JSF Error handler

**Reply by nimo23** [reply]
>I am pretty sure that is the issue is MyFaces Servlet is intercepting the requests BEFORE Quinoa gets them.

If it's not a bug on `quarkus-myfaces`, then it would be good if we could specify an exclusion pattern, so that `quarkus-myfaces` knows what to ignore, for example:

```
// with this setting, quarkus-faces does not handle html files 
// (But I don't know how Faces Servlet handles CSS, JS, etc. 
// that should belong to the Quinoa webapp..
quarkus.myfaces.ignore.pattern=*.html
```
or 

```
// quarkus-faces does not handle content within the react directory
quarkus.myfaces.ignore.dir=/react
```

A better solution would be to not need such settings. I think `quarkus-myfaces` is doing a switch in `prod`-mode that results in this behavior. Maybe we can switch it back manually. But what settings are they and is it _safe_ to switch it back?

**Reply by melloware** [reply]
It could also be a VertX ordering where in Dev Mode VertX orders Quinoa higher and in Prod Mode its ordering MyFaces higher.

**Reply by nimo23** [reply]
>It could also be a VertX ordering where in Dev Mode VertX orders Quinoa higher and in Prod Mode its ordering MyFaces higher.

Yes, I think so too. For example, with the following setting in `application.properties`:

`quarkus.swagger-ui.always-include=true`

the html pages of the swagger-ui can also be accessed in `prod`-mode (i.e. no faces servlet, which has priority here).

**Reply by melloware** [reply]
might need @ia3andy to chime in here.

**Comment by nimo23** [comment, marked as answer]
@melloware Using JSF and Quinoa in the same project works! We just need to set up a dedicated path for Quinoa so that JSF doesn't interfere with the following in `application.properties`:

`quarkus.quinoa.ui-root-path=/quinoa`

The following reproducer works for `dev` and `prod` mode:

[quarkus-reproducer.zip](https://github.com/user-attachments/files/21574906/quarkus-reproducer.zip)

- Link for Jakarta Faces: `localhost:8080/index.xhtml` or `localhost:8080`
- Link for Quinoa Vite: `localhost:8080/quinoa`

**Reply by melloware** [reply]
@nimo23 just curious for my knowledge does the JSF app talk to the Quinoa app or vice versa?  Or are they two completely unrelated apps?  I keep getting asked by my one client about having a JSF app have a menu option that goes to a React App and vice versa the React app being able to link to functions on the JSF app?  I am just curious of your experience??

**Reply by nimo23** [reply]
>Or are they two completely unrelated apps?

Yes, they are two independent/isolated frontends. 

You can link from react-frontend to jsf-frontend simply by using, e.g.

`<a href='/index.xhtml' target="_blank">Go to Jakarta Faces Version</a>`

You can link from jsf-frontend to react-frontend simply by using, e.g.

`<p:linkButton href="/quinoa" target="_blank" value="Go to React Version" icon="pi pi-external-link" styleClass="ui-button-outlined ui-button-secondary" />`

You could pass arguments between react and jsf by using URL query params from href, e.g. `href="/quinoa?filter=1"`).


Check out the updated reproducer – it includes these links:

[quarkus-reproducer.zip](https://github.com/user-attachments/files/21576664/quarkus-reproducer.zip)
